lib/state/simperium/middleware.ts

```diff
@@ -9,48 +11,88 @@ import * as T from '../../types';
 
 const debug = debugFactory('simperium-middleware');
 
-type Buckets = {
-  note: T.Bucket<T.Note>;
-};
-
-const buckets: Buckets = {} as Buckets;
+export const initSimperium = (
+  logout: () => any,
+  token: string,
+  username: string | null,
+  createWelcomeNote: boolean,
+  client: Client<'note' | 'preferences' | 'tag'>
+): S.Middleware => (store) => {
+  client.on('message', (message: string) => {
+    if (!message.startsWith('0:auth:')) {
+      return;
+    }
 
-export const storeBuckets = (newBuckets: Buckets) => {
-  buckets.note = newBuckets.note;
-};
+    const [prefix, authenticatedUsername] = message.split('0:auth:');
+    debug(`authenticated: ${authenticatedUsername}`);
 
-const fetchRevisions = (store: S.Store, state: S.State) => {
-  if (!state.ui.showRevisions || !state.ui.note) {
-    return;
+    if (username !== authenticatedUsername) {
+      debug(`was logged in as ${username} - logging out`);
+      return logout();
+    }
+  });
+
+  client.on('unauthorized', () => {
+    logout();
+  });
+
+  const noteBucket = client.bucket('note');
+
+  if (createWelcomeNote) {
+    import(
+      /* webpackChunkName: 'welcome-message' */ '../../welcome-message'
+    ).then(({ content }) => {
+      const now = Date.now() / 1000;
+      noteBucket.add({
+        content,
+        deleted: false,
+        systemTags: [],
+        creationDate: now,
+        modificationDate: now,
+        shareURL: '',
+        publishURL: '',
+        tags: [],
+      });
+    });
   }
 
-  const note = state.ui.note;
+  const fetchRevisions = (store: S.Store, state: S.State) => {
+    if (!state.ui.showRevisions || !state.ui.note) {
+      return;
+    }
 
-  buckets.note.getRevisions(
-    note.id,
-    (error: unknown, revisions: T.NoteEntity[]) => {
-      if (error) {
-        return debug(`Failed to load revisions for note ${note.id}: ${error}`);
-      }
+    const note = state.ui.note;
 
-      const thisState = store.getState();
-      if (!(thisState.ui.note && note.id === thisState.ui.note.id)) {
-        return;
-      }
+    noteBucket.getRevisions(
+      note.id,
+      (error: unknown, revisions: T.NoteEntity[]) => {
+        if (error) {
+          return debug(
+            `Failed to load revisions for note ${note.id}: ${error}`
+          );
+        }
 
-      store.dispatch(actions.ui.storeRevisions(note.id, revisions));
-    }
-  );
-};
+        const thisState = store.getState();
+        if (!(thisState.ui.note && note.id === thisState.ui.note.id)) {
```

Comment:
why use the parens with the bang?

`thisState.ui.note && note.id !== thisState.ui.note.id`

Reply:
hm. I think you are asking why I didn't write different logic? as written, the parentheses is important because the `!` takes precedence in associativity over the `&&`

the logic you wrote in the comment _is_ an alternative to the `!` but it fails to grab the case when `thisState.ui.note` is `null` - the condition would fail and we would continue to dispatch the revisions for a note that's not longer selected.

whenever we want to invert logic we need to rely on deMorgan's laws for the full flip, otherwise we'd miss cases like this. `&&` and `||` swap and every term has to be negated.

```ts
if (!(thisState.ui.note && note.id === thisState.ui.note.id)) {}
if (!thisState.ui.note || note.id !== thisState.ui.note.id) {}
```

at this point I often prefer to write the first case since it reads easier for me. the difference is "if we don't have: a note and one which matches the id" vs. "if we don't have a note or we have a note but the id doesn't match." this tends to be more notable as the list gets larger with more terms but in general I do find the symmetry of this construction to be helpful. "look for all the requirements and _then_ if we don't have all of them, take this other route"